File: backend/app/services/feed.py
"""Knowledge Feed assembly: turn the user's saved cards into a shuffled stream of
bite-sized "moments" — the reel format, but for their own knowledge.

Moment kinds:
  - insight     : the card's one-liner / TL;DR (the core takeaway)
  - highlight   : a punchy line pulled from the card body
  - quiz        : a single stored quiz question (interactive)
  - thread      : a rabbit-hole thread the reader can fall into
  - connection  : a serendipitous link between two cards (see serendipity.py)

Everything except `connection` is built from data ALREADY stored on the card —
zero LLM cost. Connections reuse the cached serendipity links (and top up a small,
bounded number). Moments are interleaved so consecutive items rarely share a card.
"""

from __future__ import annotations

import logging
import random
from collections import deque

from app.services import serendipity
from app.store import media as media_store

log = logging.getLogger("services.feed")

_MAX_THREADS_PER_CARD = 2
_MAX_QUIZ_PER_CARD = 2
_CONNECTIONS_WANT = 4
_CONNECTIONS_MAX_NEW = 2  # bound the LLM cost of a feed load


def _card_ref(row) -> dict:
    return {
        "card_id": row.id,
        "title": (row.one_liner or row.caption or "Untitled").strip()[:120],
        "content_type": row.content_type or "other",
        "thumbnail": media_store.to_media_url(row.thumbnail),
    }


def _clean(text: str, limit: int = 220) -> str:
    return " ".join((text or "").split()).strip()[:limit]


def _best_highlight(blocks: list) -> str | None:
    """Pull one punchy line from the card body — a callout, else a bullet, else a
    paragraph. Returns None when there's nothing quotable."""
    callout = bullet = paragraph = None
    for b in blocks or []:
        if not isinstance(b, dict):
            continue
        btype = b.get("type")
        if btype == "callout" and callout is None and b.get("text"):
            callout = _clean(b["text"])
        elif btype == "bullet_list" and bullet is None:
            items = [i for i in (b.get("items") or []) if isinstance(i, str) and i.strip()]
            if items:
                bullet = _clean(items[0])
        elif btype == "paragraph" and paragraph is None and b.get("text"):
            paragraph = _clean(b["text"])
    return callout or bullet or paragraph


def _moments_for_card(row) -> list[dict]:
    ref = _card_ref(row)
    cid = row.id
    out: list[dict] = []

    tldr = _clean(row.tldr or row.one_liner, limit=280)
    if tldr:
        out.append({"id": f"{cid}:insight", "kind": "insight", "card": ref, "text": tldr})

    highlight = _best_highlight(row.blocks)
    # Skip a highlight that just echoes the insight text.
    if highlight and highlight.lower() != tldr.lower():
        out.append({"id": f"{cid}:highlight", "kind": "highlight", "card": ref, "text": highlight})

    insight = row.insight if isinstance(row.insight, dict) else {}

    quiz = insight.get("quiz") if isinstance(insight.get("quiz"), list) else []
    for k, q in enumerate(quiz[:_MAX_QUIZ_PER_CARD]):
        if not isinstance(q, dict):
            continue
        options = [str(o) for o in (q.get("options") or [])]
        idx = q.get("answer_index")
        question = _clean(q.get("question", ""), limit=200)
        if question and 2 <= len(options) <= 4 and isinstance(idx, int) and 0 <= idx < len(options):
            out.append({
                "id": f"{cid}:quiz:{k}",
                "kind": "quiz",
                "card": ref,
                "question": question,
                "options": options,
                "answer_index": idx,
                "explanation": _clean(q.get("explanation", ""), limit=240),
            })

    rh = insight.get("rabbit_hole") if isinstance(insight.get("rabbit_hole"), dict) else {}
    threads: list[str] = []
    for key in ("questions", "adjacent_topics", "advanced_concepts"):
        for t in (rh.get(key) or []):
            if isinstance(t, str) and t.strip():
                threads.append(t.strip())
    for k, t in enumerate(threads[:_MAX_THREADS_PER_CARD]):
        out.append({"id": f"{cid}:thread:{k}", "kind": "thread", "card": ref, "text": _clean(t, limit=160)})

    return out


def _interleave(moments: list[dict]) -> list[dict]:
    """Round-robin across per-card buckets so consecutive moments rarely share a
    card. Connections each get their own bucket, so they scatter naturally."""
    buckets: dict[str, deque] = {}
    for m in moments:
        key = m["id"] if m["kind"] == "connection" else m["card"]["card_id"]
        buckets.setdefault(key, deque()).append(m)
    queues = list(buckets.values())
    random.shuffle(queues)
    out: list[dict] = []
    while queues:
        for q in queues:
            if q:
                out.append(q.popleft())
        queues = [q for q in queues if q]
    return out


async def build_feed(
    session, *, owner_id: str | None, cards: list, limit: int
) -> list[dict]:
    """Assemble the owner's feed. `cards` are READY CardRows (owner-scoped)."""
    moments: list[dict] = []
    for row in cards:
        moments.extend(_moments_for_card(row))

    # Weave in serendipitous connections (cheap: cached + a couple fresh).
    try:
        conns = await serendipity.get_connections(
            session,
            owner_id=owner_id,
            cards=cards,
            want=_CONNECTIONS_WANT,
            max_new=_CONNECTIONS_MAX_NEW,
        )
    except Exception as e:  # noqa: BLE001 — connections are a bonus, never fatal
        log.warning("feed: connection generation failed: %s", e)
        conns = []
    for c in conns:
        moments.append({
            "id": f"conn:{c['card_a']['card_id']}:{c['card_b']['card_id']}",
            "kind": "connection",
            "card": c["card_a"],
            "card_b": c["card_b"],
            "text": c["blurb"],
        })

    random.shuffle(moments)
    return _interleave(moments)[:limit]

File: app/lib/domain/models/feed.dart
/// Models for the Knowledge Feed and the Connections (serendipity) view — a
/// reel-style stream of "moments" built from the user's own cards, plus the
/// surprising links the backend finds between pairs of cards. Mirrors the
/// backend schemas in api/feed.py and api/connections.py.
library;

import 'enums.dart';

/// A lightweight pointer to a source card, enough to render a moment header and
/// deep-link into the reader.
class FeedCardRef {
  const FeedCardRef({
    required this.cardId,
    required this.title,
    required this.contentType,
    this.thumbnail,
  });

  final String cardId;
  final String title;
  final ContentType contentType;
  final String? thumbnail;

  factory FeedCardRef.fromJson(Map<String, dynamic> json) => FeedCardRef(
        cardId: (json['card_id'] as String?) ?? '',
        title: (json['title'] as String?) ?? '',
        contentType: ContentType.fromWire(json['content_type'] as String?),
        thumbnail: json['thumbnail'] as String?,
      );
}

enum FeedItemKind {
  insight,
  highlight,
  quiz,
  thread,
  connection,
  unknown;

  static FeedItemKind fromWire(String? value) {
    switch (value) {
      case 'insight':
        return FeedItemKind.insight;
      case 'highlight':
        return FeedItemKind.highlight;
      case 'quiz':
        return FeedItemKind.quiz;
      case 'thread':
        return FeedItemKind.thread;
      case 'connection':
        return FeedItemKind.connection;
      default:
        return FeedItemKind.unknown;
    }
  }
}

/// One swipeable moment in the feed. Fields are kind-specific; unused ones stay
/// at their defaults.
class FeedItem {
  const FeedItem({
    required this.id,
    required this.kind,
    required this.card,
    this.text = '',
    this.question = '',
    this.options = const [],
    this.answerIndex = 0,
    this.explanation = '',
    this.cardB,
  });

  final String id;
  final FeedItemKind kind;
  final FeedCardRef card;

  /// insight TL;DR / highlight line / thread prompt / connection blurb.
  final String text;

  // quiz-specific
  final String question;
  final List<String> options;
  final int answerIndex;
  final String explanation;

  /// connection-specific — the second card in the pair.
  final FeedCardRef? cardB;

  bool get quizValid =>
      question.trim().isNotEmpty &&
      options.length >= 2 &&
      answerIndex >= 0 &&
      answerIndex < options.length;

  factory FeedItem.fromJson(Map<String, dynamic> json) => FeedItem(
        id: (json['id'] as String?) ?? '',
        kind: FeedItemKind.fromWire(json['kind'] as String?),
        card: FeedCardRef.fromJson(
          (json['card'] as Map<String, dynamic>?) ?? const {},
        ),
        text: (json['text'] as String?) ?? '',
        question: (json['question'] as String?) ?? '',
        options: ((json['options'] as List?) ?? const [])
            .map((e) => e.toString())
            .toList(),
        answerIndex: (json['answer_index'] as num?)?.toInt() ?? 0,
        explanation: (json['explanation'] as String?) ?? '',
        cardB: json['card_b'] is Map<String, dynamic>
            ? FeedCardRef.fromJson(json['card_b'] as Map<String, dynamic>)
            : null,
      );
}

/// A surprising link between two of the user's cards (the serendipity engine).
class Connection {
  const Connection({
    required this.cardA,
    required this.cardB,
    required this.blurb,
  });

  final FeedCardRef cardA;
  final FeedCardRef cardB;
  final String blurb;

  factory Connection.fromJson(Map<String, dynamic> json) => Connection(
        cardA: FeedCardRef.fromJson(
          (json['card_a'] as Map<String, dynamic>?) ?? const {},
        ),
        cardB: FeedCardRef.fromJson(
          (json['card_b'] as Map<String, dynamic>?) ?? const {},
        ),
        blurb: (json['blurb'] as String?) ?? '',
      );
}
